lib/ipam/ipam.go

```diff
@@ -854,17 +876,39 @@ func (c ipamClient) ReleaseIPs(ctx context.Context, ips []net.IP) ([]net.IP, err
 		ipsByBlock[cidrStr] = append(ipsByBlock[cidrStr], ip)
 	}
 
-	// Release IPs for each block.
+	// Release IPs for each block. These don't typically compete for resources, so we can do them in parallel
+	// in order to move quickly.
+	type retVal struct {
+		Error       error
+		Unallocated []net.IP
+	}
+	resultChan := make(chan retVal, len(ipsByBlock))
+	log.Infof("Starting %d routines to release IP addresses", len(ipsByBlock))
 	for cidrStr, ips := range ipsByBlock {
 		_, cidr, _ := net.ParseCIDR(cidrStr)
-		unalloc, err := c.releaseIPsFromBlock(ctx, ips, *cidr)
-		if err != nil {
-			log.Errorf("Error releasing IPs: %v", err)
-			return nil, err
+		go func(cidr net.IPNet, ips []net.IP) {
+			r := retVal{}
+			unalloc, err := c.releaseIPsFromBlock(ctx, ips, cidr)
+			if err != nil {
+				log.Errorf("Error releasing IPs: %v", err)
+				r.Error = err
+			}
+			r.Unallocated = unalloc
+			resultChan <- r
+		}(*cidr, ips)
+	}
+
+	// Read the response from each goroutine.
+	err = nil
+	for i := 0; i < len(ipsByBlock); i++ {
+		r := <-resultChan
+		log.Debugf("Received response #%d from release goroutine", i)
+		if r.Error != nil && err == nil {
+			err = r.Error
```

Comment:
Should we have a way to join all the possible errors together? Seems like we would want to report all of the IPs we attempted to release.

Reply:
Maybe. I figured that we do log out every error as we hit them, and this is just to report back that _some_ error occurred.

We might be able to return a []error? But I'm not sure what clients would do with that vs. just a single error.

Reply:
My read of this was we could hit a bunch of errors but we would only be reporting the last one that we receive? Maybe we just get around this by just logging a warning and then return back the last error that we get as per usual?

Reply:
Yep, we log an error up on line 892 already